### PR TITLE
FEAT: New mech module: Engineering toolset! Also huge QoL for all mech users

### PR DESCRIPTION
**Pull request by ErdGinalD**
## What Does This PR Do
Adds a toolbox module to the game containing a multitool, screwdriver, wrench, crowbar, and welder. All items have a 0.5 speed modifier. When emagged, it unlocks an extremely dangerous knife for 25 damage and 20 flat penetration, effectively providing an alternative to the mecha claw, which can kill silently due to oxy damage.  

Modules are selected by pressing the Z button, which opens a radial selection menu. 

Action buttons are now available for all mech-users, allowing them to bind module switching and interact with the mech faster in general.

## Why It's Good For The Game
I think it's a good idea to give more life to engineering mechs by letting them fix wiring, doors, and blatantly steal NanoMeds from the med bay without getting out of the mech. At the same time allows the Malf AI to play less dependent on the Borg, with the help of mechs, building their own line of defense or otherwise implementing a new and definitely needed tool.

Action buttons, on the other hand, despite having an Atl+click function, are more intuitive and allow you to bind modules, fighting or fixing a station more efficiently, switching between RDC or a new toolbox etc. A useful and cool feature, proven over time on Russian Paradise, which never once gave the question "what's it for".

## Images of changes
image
Act like green toolbox(not His Grace) in hand, same visual and sound
image
Radial menu
image
multitool interaction
image
Action button
image
image

## Testing
Code, Testing, Pray, almost give up but again rize up. Now it work perfectly as i can see it.

## Changelog
:cl:
add: New nodule for mecha: Toolset module
add: Action buttons for mecha modules now online
/:cl:

### CODE REVIEW

File: code/_onclick/item_attack.dm [hunk 1]
```diff
@@ -66,12 +66,12 @@
 		return
 
 	if(!force)
-		playsound(loc, 'sound/weapons/tap.ogg', get_clamped_volume(), 1, -1)
+		playsound(user.loc, 'sound/weapons/tap.ogg', get_clamped_volume(), 1, -1)
```
Comment (Burzah) [inline code/_onclick/item_attack.dm:69]:
```suggestion
		playsound(user.loc, 'sound/weapons/tap.ogg', get_clamped_volume(), TRUE, -1)
```

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 2]
```diff
@@ -278,6 +278,8 @@
 					playsound(target, usesound, 50, 1)
 					playsound(target, 'sound/effects/sparks2.ogg', 50, 1)
```
Comment (Burzah) [inline code/game/mecha/equipment/tools/work_tools.dm:279]:
```suggestion
					playsound(target, usesound, 50, TRUE)
					playsound(target, 'sound/effects/sparks2.ogg', 50, TRUE)
```
Can you fix these while we are here?

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 3]
```diff
@@ -278,6 +278,8 @@
...
 
+/obj/item/mecha_parts/mecha_equipment/rcd/proc/check_menu(mob/living/carbon/user)
+	return (user && chassis.occupant == user && user.stat != DEAD)
 
 /obj/item/mecha_parts/mecha_equipment/rcd/Topic(href,href_list)
```
Comment (Burzah) [inline code/game/mecha/equipment/tools/work_tools.dm:284]:
```suggestion
/obj/item/mecha_parts/mecha_equipment/rcd/Topic(href, href_list)
```
Another one, while we are here.

File: code/game/mecha/mecha_actions.dm [hunk 4]
```diff
@@ -231,3 +236,24 @@
 	button_icon_state = "mech_damtype_[new_damtype]"
 	playsound(src, 'sound/mecha/mechmove01.ogg', 50, 1)
```
Comment (Burzah) [inline code/game/mecha/mecha_actions.dm:237]:
```suggestion
	playsound(src, 'sound/mecha/mechmove01.ogg', 50, TRUE)
```
While we are here.

**Review by Burzah — CHANGES_REQUESTED** [review]

File: code/modules/research/designs/mechfabricator_designs.dm [hunk 5]
```diff
@@ -633,6 +633,17 @@
...
 	category = list("Exosuit Equipment")
 
+/datum/design/mech_eng_toolset
+	name = "Exosuit Engineering Equipment (Engineering Toolset)"
+	desc = "Exosuit toolset. Gives a set of good tools."
+	id = "mech_eng_toolset"
+	build_type = MECHFAB
+	req_tech = list("materials" = 6, "engineering" = 4)
+	build_path = /obj/item/mecha_parts/mecha_equipment/eng_toolset
+	materials = list(MAT_METAL=10000,MAT_TITANIUM =2000,MAT_PLASMA=2000)
```
Comment (Sirryan2002) [inline code/modules/research/designs/mechfabricator_designs.dm:643]:
```suggestion
	materials = list(MAT_METAL = 10000, MAT_TITANIUM = 2000, MAT_PLASMA = 2000)
```

File: code/game/mecha/mecha_actions.dm [hunk 6]
```diff
@@ -231,3 +236,24 @@
...
 	UpdateButtonIcon()
+
+/datum/action/innate/mecha/select_module
+	name = "Hey, you shouldn't see it"
```
Comment (Sirryan2002) [inline code/game/mecha/mecha_actions.dm:241]:
```suggestion
	name = "Hey, you shouldn't see this"
```

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 7]
```diff
@@ -455,6 +457,96 @@
...
+		selected_item = selected
+		occupant_message("Switched to [selected_item]")
+		update_equip_info()
+
+/obj/item/mecha_parts/mecha_equipment/eng_toolset/emag_act(mob/user)
+	if(!emagged)
+		items_list.Add(new emag_item)
+		emagged = TRUE
+		user.visible_message("<span class='warning'>Sparks fly out of [src.name]!</span>", "<span class='notice'>You short out the safeties on [src.name].</span>")
+		playsound(src.loc, 'sound/effects/sparks4.ogg', 50, TRUE)
```
Comment (Sirryan2002) [inline code/game/mecha/equipment/tools/work_tools.dm:547]:
```suggestion
		playsound(loc, 'sound/effects/sparks4.ogg', 50, TRUE)
```

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 8]
```diff
@@ -455,6 +457,96 @@
...
+	if(selected in items_list)
+		selected_item = selected
+		occupant_message("Switched to [selected_item]")
+		update_equip_info()
+
+/obj/item/mecha_parts/mecha_equipment/eng_toolset/emag_act(mob/user)
+	if(!emagged)
+		items_list.Add(new emag_item)
+		emagged = TRUE
+		user.visible_message("<span class='warning'>Sparks fly out of [src.name]!</span>", "<span class='notice'>You short out the safeties on [src.name].</span>")
```
Comment (Sirryan2002) [inline code/game/mecha/equipment/tools/work_tools.dm:546]:
```suggestion
		user.visible_message("<span class='warning'>Sparks fly out of [name]!</span>", "<span class='notice'>You short out the safeties on [name].</span>")
```

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 9]
```diff
@@ -455,6 +457,96 @@
...
+		else
+			. += "|<a href='?src=[UID()];select=[item.UID()]'>[short_name]</a>"
+	. += "|"
+
+/obj/item/mecha_parts/mecha_equipment/eng_toolset/Topic(href,href_list)
+	..()
+	if(href_list["select"])
+		selected_item = locateUID(href_list["select"])
+		occupant_message("Switched to [selected_item]")
+		update_equip_info()
```
Comment (Sirryan2002) [inline code/game/mecha/equipment/tools/work_tools.dm:504]: you need to sanity check this UID. If you get a null href value or the item somehow gets deleted, you need to make sure that the toolset doesn't break.

**Review by Sirryan2002 — CHANGES_REQUESTED** [review]

File: code/game/mecha/equipment/tools/work_tools.dm (hunk 9 above)
Comment (ErdGinalD) [reply code/game/mecha/equipment/tools/work_tools.dm:504]: How to properly do that? 
else return?

Comment (Sirryan2002) [reply code/game/mecha/equipment/tools/work_tools.dm:504]:
```dm
if(!istype(selected_item)
   return
```

Comment (ErdGinalD) [reply code/game/mecha/equipment/tools/work_tools.dm:504]: Update it, thanks!

**Comment by Burzah** [comment]
CI has failed because of spacing/indent issues on line 506 of `work_tools.dm`. This will need to be fixed for it to pass testing.

File: code/game/mecha/equipment/tools/work_tools.dm (hunk 9 above)
Comment (ErdGinalD) [reply code/game/mecha/equipment/tools/work_tools.dm:504]: > ```dm
> if(!istype(selected_item)
>    return
> ```

Not working, i dunno how to do it properly but your variant is not working

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 10]
```diff
@@ -455,6 +457,96 @@
 	last_piece = NC
 	return TRUE
 
+/obj/item/mecha_parts/mecha_equipment/eng_toolset
+	name = "Engineering toolset"
```
Comment (Henri215) [inline code/game/mecha/equipment/tools/work_tools.dm:461]:
```suggestion
	name = "engineering toolset"
```

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 11]
```diff
@@ -455,6 +457,96 @@
...
+			. += "|<b>[short_name]</b> "
+		else
+			. += "|<a href='?src=[UID()];select=[item.UID()]'>[short_name]</a>"
+	. += "|"
+
+/obj/item/mecha_parts/mecha_equipment/eng_toolset/Topic(href,href_list)
+	..()
+	if(href_list["select"])
+		selected_item = locateUID(href_list["select"])
+		occupant_message("Switched to [selected_item]")
```
Comment (Henri215) [inline code/game/mecha/equipment/tools/work_tools.dm:503]:
```suggestion
		occupant_message("<span class='notice'>Switched to [selected_item].</span>")
```

File: code/game/mecha/equipment/tools/work_tools.dm [hunk 12]
```diff
@@ -455,6 +457,96 @@
...
+		if(item.name == choice)
+			selected = item
+			break
+	if(selected)
+		extend(selected)
+
+/obj/item/mecha_parts/mecha_equipment/eng_toolset/proc/extend(obj/item/selected)
+	if(selected in items_list)
+		selected_item = selected
+		occupant_message("Switched to [selected_item]")
```
Comment (Henri215) [inline code/game/mecha/equipment/tools/work_tools.dm:539]:
```suggestion
		occupant_message("<span class='notice'>Switched to [selected_item].</span>")
```

File: code/game/mecha/equipment/tools/work_tools.dm (hunk 9 above)
Comment (Sirryan2002) [reply code/game/mecha/equipment/tools/work_tools.dm:504]: is it properly indented?

**Comment by Exavere** [comment]
Sup gamer, just a few questions I've garnered from some nerds' concerns.  

Is the Welding Tool within the Module infinite?

This isn't limited, and can be applied to any of the Mecha's atm right?

I assume mechs will have the ability to repair themselves, and others?

If these are all true, there may be a bit of an issue when you consider two combat mechas fighting together; or even alone when they can fix themselves non-stop.

**Comment by ErdGinalD** [comment]
> Sup gamer, just a couple questions I've garnered from some nerds' concerns.
> 
> Is the Welding Tool within the Module infinite?
> 
> This isn't limited to RIPLEYs, and can be applied to any of the Mecha's right?

Welding tool is inlimite and can be attached to RIPLEY and Fifefighter only. Will fix all PR in a week. got too busy

File: code/game/mecha/equipment/tools/work_tools.dm (hunk 9 above)
Comment (ErdGinalD) [reply code/game/mecha/equipment/tools/work_tools.dm]: Done

**Comment by ErdGinalD** [comment]
> CI has failed because of spacing/indent issues on line 506 of `work_tools.dm`. This will need to be fixed for it to pass testing.

Done

File: code/modules/mob/mob.dm [hunk 13]
```diff
@@ -673,6 +673,12 @@ GLOBAL_LIST_INIT(slot_equipment_priority, list( \
 
 	DEFAULT_QUEUE_OR_CALL_VERB(VERB_CALLBACK(src, PROC_REF(run_mode)))
 
+	if(ismecha(loc))
+		var/obj/mecha/mecha = loc
+		if(src == mecha.occupant)
+			mecha.selected?.self_occupant_attack()
+		return
```
Comment (DGamerL) [inline code/modules/mob/mob.dm:680]:
```suggestion
```
Implicit return

**Pull request closed without merging**